Enhancement/29 support keyboard binds for pull requests

hk0i opened the pull request:
Added support for key bindings to specific events in bit bucket pull requests.
# Default Key Mapping
## Tab Selection

![bitbucket pull request tabs](https://cloud.githubusercontent.com/assets/755669/18594922/1c74c184-7bf7-11e6-887d-859fb6206c65.png)

| Key | Action |
| --- | --- |
| <kbd>1</kbd> | Selects the **Overview** tab |
| <kbd>2</kbd> | Selects the **Commits** tab |
| <kbd>3</kbd> | Selects the **Activity** tab |
## Page Navigation and Scrolling

_Upper case letters mean that you need to press shift and the letter_

| Key | Action |
| --- | --- |
| <kbd>N</kbd> | Scrolls to the **N**ext comment on the page. |
| <kbd>P</kbd> | Scrolls to the **P**previous comment on the page. |
| <kbd>g</kbd> <kbd>g</kbd> | Scrolls to the top of the page |
| <kbd>G</kbd> | Scrolls to the bottom of the page |

hk0i committed:
(29-support-keyboard-binds) first working commit

- Added the code from my greasemonkey script into code base
- Wired up events
- Sending mouse click to DOM element instead of jQuery object.
- Wrapped next and previous comment binds with a call to
  `waitForRender()`
- Added module.exports stuffs

hk0i committed:
(29-support-keyboard-binds) added comments

- Added some doc comments
- Added some code that may allow us to add customizable keymaps in the
  future to override the default keymap.
- Switched some functions to use ES6 arrow function syntax, since I saw
  it elsewhere in the code.

hk0i committed:
added contributors and updated readme

- Added notes to install browserify in readme
- Started CONTRIBUTORS.md file for contributes, cause hey why not?
  Add your name if you contributed!

hk0i committed:
(29-support-keyboard-binds) fixed broken build (oops)

- Fixed broken build due to some incomplete code that must've been deleted
  by mistake

- Added comment jumping compatibility with <kbd>n</kbd> and <kbd>p</kbd>
  to the **Activity** tab.

- Optimized the creation of the comments array so we don't always
  create a new one when you press n or p.

- Keybinds are made indefinitely instead of waiting for the element to
  render.

- Added jshint suppression at the top of the file to eliminate ES6
  warnings.

hk0i committed:
(29-support-keyboard-binds) removed jQuery ready from main

- Removed jQuery ready event from main. This was originally added to
  try and address an issue with calling the `click()` method of a
  jQuery object, but it had no effect and everything works without it.

hk0i committed:
(29-support-keyboard-binds) added documentation

- Added default keybinds to `README.md`

andremw commented on CONTRIBUTORS.md:
```diff
@@ -0,0 +1,3 @@
+# Contributors
+
+[Gregory McQuillan](http://github.com/hk0i)
```
Here you can also add Adam Pitzele @apitzele, Rodrigo Proença @rproenca and me.

hk0i replied:
sure. should we all go under contributors or should we have a separate `AUTHORS` file for original authors? Either way I don't mind the extra company

andremw replied:
Ah, let's put all the contributors in one section. :)

hk0i replied:
Done.

andremw commented on README.md:
````diff
@@ -10,19 +10,67 @@ and made a few improvements. Kudos!
+If you don't have it already, you will need [Browserify][browserify]. It is available on npm:
+
+```
+npm install -g browserify
````
You can also install it with npm install. It's listed as a devDependency.

hk0i replied:
ah okay I missed that :)

hk0i replied:
updated readme

andremw commented on README.md:
````diff
@@ -10,19 +10,67 @@ and made a few improvements. Kudos!
+If you don't have it already, you will need [Browserify][browserify]. It is available on npm:
+
+```
+npm install -g browserify
````
You can actually install it with npm install. It's listed as a devDependency.

andremw commented on src/keymap/keymap.js:
```diff
@@ -0,0 +1,184 @@
+     *
+     * Waits for `selector` to be available before initializing comments.
+     *
+     * @param {String} selector waits for this selecor to become available
```
typo here: waits for this _selecor_

hk0i replied:
fixed

andremw commented on src/keymap/keymap.js:
```diff
@@ -0,0 +1,184 @@
+    };
+
+    /**
+     * Scrolls the browser window to the previous comment on the PR diff.
```
I think it should be "Scrolls the browser window to the _next_ comment.."

hk0i replied:
good catch. Fixed.

andremw commented on src/keymap/keymap.js:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        keyboard.reset();
+        self.comments = document.querySelectorAll(self.commentSelector);
```
Here I think we should use waitForRender because the diff section renders after everything else on the screen and the only rendered comments when the page loads are the PR comments and not the diff comments.

andremw replied:
And I wonder if it's not main.js's responsability to control the need for this waitForRender. The keymap should not know about these things.

hk0i replied:
You're right, that one slipped out unnoticed. I moved it around a bunch trying to get it to work in the original code and this must've been left behind.

I'm not sure about the waitForRender... Would the waitForRender then trigger the entire keymap? The only keys we need to wait for rending on are the comment keys, the tab changing and scroll to top and bottom should still work. The selector we are waiting for also changes depending on the tab we have activated and will need to wait again when we switch back and forth between _Overview_ and _Activity_ tabs.

andremw replied:
You're right. There's no need to complicate here. Let's use the waitForRender in this line only.

hk0i replied:
This won't work (I did try it just to make sure).
I _did_ swap out the call for `document.querySelectorAll()` and I changed it to use `self.initComments()`, however it breaks on tab switching. 

I originally had extra calls when they use the keyboard to switch tabs, but I just realized that also doesn't handle if they click those buttons with the mouse. I've added a couple of click listeners to handle the tab changing.

If there's a better way you can think of handling this let me know. It works now with click or key bind.

andremw replied:
I think it's good!

andremw reviewed (changes requested):
I spotted some small things that should be improved. I would also ask you to take a look at the things code climate and travis ci are complaining about.
However, I think you did a great job here. I tested the extension and it works like a charm. 
I liked the changes you made to README.md too. Can you add to it the things related to linting configuration (to avoid the contributors missing it and having travis ci complaining about the code)?

hk0i reviewed (commented):
Fixed most things except for the travis and linter issues. Gonna tackle those next.

hk0i committed:
(29-support-keyboard-binds) initial code review changes

- Added everyone else to CONTRIBUTORS.md.
- Updated instructions to install dependencies in README.md.
- Fixed some minor typos in keymap.js
- Now making a call to `initComments()` instead of manual assignment in
  the `init()` function.

hk0i committed:
(29-support-keyboard) code review changes

- Fixed a bug where comment navigation would break if the user changes
  tabs using the mouse instead of the keyboard shortcuts.

hk0i commented:
Oh geez I didn't realize the push would kick off CI again... I'm working on that part now :)

hk0i committed:
(29-support-keyboard-binds) first round of lint changes

- First round of linter changes for code style compliance, most
  everything is done except camel case changes.

hk0i committed:
(20-support-keyboard-binds) addressed camel case lint

- Switched variables to use camel case or SCREAMING_SNAKE_CASE to avoid
  issues with `npm test`.

hk0i committed:
(29-support-keyboard-binds) updated CONTRIBUTORS and README.

- Updated `README.md` to show details for running tests.
- Updated `CONTRIBUTORS.md` to use a list and provide details about
  adding and maintaining the list order.

hk0i commented:
Code climate and travis have conflicting opinions on what to do about IFFEs, so they won't both pass for my code, it will always be one or the other.

code climate wants the definitions like

``` javascript
(function ($) { }(jQuery));
```

Travis (`npm test`) wants it like this:

``` javascript
(function ($) { })(jQuery);
```

... I personally prefer the latter and since `npm test` is what's going to run on most people's machines (_is it possible to run the same lint as on code climate?_) I think it makes sense to stick with that and update the rules on code climate. But it's ultimately your decision.

andremw commented:
No problem, I'll ignore Code Climate for now, and then I'll disable this check there.

apitzele commented:
this is  a very useful addition and I will absolutely be using it.  Thanks!

I noticed a small typo in the readme.   "scrolls to the Ppreviou"  Sorry, I would have liked to give this a proper review this weekend but I wasn't around my lappy.
No biggie but maybe something that can get corrected on the next PR that goes in.

I did pull the code down and verified the keymaps work in Windows.  Looks great.
👍

hk0i commented:
Good catch on the typo, looks like it got vimmed by mistake somehow
